### ISSUE TITLE
What are the key differences between Falco and tetragon?

### ISSUE THREAD

**Issue by rkdutta** [cilium/tetragon#701]
I am trying to understand the differences between Falco and tetragon. Let’s say, falco is meant for this and it cannot do this like tetragon. Can I get some help? Kindly try to provide elaborate answers for comprehension. They are different tools but somehow I am not understanding the core differences.

**Comment by aidanhall34** [comment]
Hey, I'm not a maintainer, just looking for answers to the same question. The best comparison for the products I have gotten so far is:
https://ebpf.io/applications/

If I understand it correctly:
Falco uses ebpf to watch syscalls, then sends violations to your output stream of choice (logs, Falco exporter etc)
Tetragon uses ebpf to watch syscalls, but is able to enforce rules (eg block X syscall etc) and alert on events.

Tldr:
Falco: watch
Tetragon: watch and enforce

If a maintainer would like to clarify that'd be appreciated

**Comment by weinong** [comment]
kubearmor doesn't require privileged container like tetragon does.

**Comment by kcq** [comment]
Both use eBPF. Tetragon is a low level eBPF toolkit that you can use to build a runtime detection and prevention system. Falco is an intrusion detection system designed to detect only and it has a lot of detection rules.

**Comment by FedeDP** [comment]
I think i am somewhat in a very good position to answer this question since i am a Falco Core maintainer and a Tetragon Approver.
Also, please note that this recent blog post has a section about that: https://cilium.io/blog/2026/01/19/tetragon-falco-migrate/ :)

Falco, as others said, is basically a system call tracer; it's event schema is fixed (syscalls exit events only nowadays, syscalls enter events were recently dropped for performance reasons); userspace adds more metadata on top, like process cache, user/group cache, and so on. Falco rule syntax build upon these events to enable writing complex rules to filter relevant events, and then sending alerts for matching ones.
Moreover, Falco supports plugins to extend its behavior outside of mere syscalls watcher.
Another relevant key difference is that Falco rules are evaluated in userspace; this means that ALL of kernel events (syscalls) have to reach userspace to allow correct evaluation; unfortunately, this is a bit limiting and event dropping it's a real problem.
During its lifetime, there were multiple attempts to improve the situation that is now better but still far from resolved.

Tetragon has a completely different view; first of all the event schema is not fixed at all; indeed `TracingPolicy` are, basically, event schema injectors (you specify with hook to attach and which data to capture).
The `TracingPolicy` selectors (ie: the Falco rule `condition`) are evaluated in the kernel; this means that the kernel takes a bit more time to execute, but at the same time events reaching userspace are already filtered and thus interesting. Moreover, this means that only a bunch of events eventually reach userspace, both giving userspace much more time to do something with them, and also lowering the problem with event drops.
As others already outlined, Tetragon can also react and enforce actions upon events, most of them in kernel space directly (eg: override function return code, kill the program, ...); this is BIG when we talk about security and not observability.
Finally, Tetragon is effectively Kubernetes aware, thanks to its operator; that is something that is currently being worked on for Falco.

These are my 2c :)
